gitlab/utils.py

```diff
@@ -47,6 +49,39 @@ def response_content(
     return None
 
 
+def _transform_types(
+    data: Dict[str, Any], custom_types: dict, transform_files: Optional[bool] = True
+) -> Tuple[dict, dict]:
+    """Copy the data dict with attributes that have custom types and transform them
+    before being sent to the server.
+
+    If ``transform_files`` is ``True`` (default), also populates the ``files`` dict for
+    FileAttribute types with tuples to prepare fields for requests' MultipartEncoder:
+    https://toolbelt.readthedocs.io/en/latest/user.html#multipart-form-data-encoder
+
+    Returns:
+        A tuple of the transformed data dict and files dict"""
+
+    # Duplicate data to avoid messing with what the user sent us
+    data = data.copy()
```

Comment:
Should this use a `copy.deepcopy()` to be safer?

Reply:
Might make sense although with this new helper we'd be calling it on most of the API calls so maybe we should check the implications (https://stackoverflow.com/questions/24756712/deepcopy-is-extremely-slow).

Seems like this code has been in create/update/list for a really long time though with no issues. Maybe I could also refactor to try to avoid running this helper every time unless needed.

Reply:
No problem keeping it as copy. If it has been working as is then it is probably fine.

I doubt there would be a noticeable performance impact from using `copy.deepcopy()`. Compared to the amount of time the network request takes. I did a simple test of doing a `copy.deepcopy()` 100,000 times and it took 1.15 seconds on my system. Which works out to 0.0000114471 seconds per copy. Now a plain `copy` was much faster at 0.028097620233893394 seconds for 100,000 iterations. Which is 0.0000002810 seconds per iteration.

But compared to how long a network request is then I feel like it will be just noise.

I was copying: `a = {"1": [1, 2, 3, 4, 5], "2": {"hello": "there"}, "3": False}`

Reply:
Ah right, we probably will be copying larger dicts than that with projects/groups, but you're right we'll probably also hit rate limits before this starts to make a difference.

Reply:
I've added deepcopy as we said IO is probably slower than this. If people complain it's slow we can revert :)

Reply:
Hmm I've had to revert the `deepcopy()` for now, because it uses pickle internally, and users might be providing some unpicklable objects in the dict:
https://github.com/python-gitlab/python-gitlab/runs/6643492246?check_suite_focus=true#step:5:282

Reply:
> Hmm I've had to revert the `deepcopy()` for now, because it uses pickle internally, and users might be providing some unpicklable objects in the dict: https://github.com/python-gitlab/python-gitlab/runs/6643492246?check_suite_focus=true#step:5:282

Makes sense. Thanks.